src/component/mongoose_component_connection.erl

```diff
@@ -306,25 +306,23 @@ try_register_routes(StateData, Retries) ->
             handle_registration_conflict(ConflictBehaviour, RoutesInfo, StateData, Retries)
     end.
 
--spec register_routes(data()) -> any().
-register_routes(StateData) ->
-    #component_data{listener_opts = #{hidden_components := AreHidden}} = StateData,
-    Routes = get_routes(StateData),
+-spec register_routes(data()) -> {ok, mongoose_component:external_component()} | {error, any()}.
+register_routes(#component_data{lserver = SubDomain,
+                                is_subdomain = true,
+                                listener_opts = #{hidden_components := AreHidden}}) ->
+    Prefix = {prefix, <<SubDomain/binary, ".">>},
     Handler = mongoose_packet_handler:new(mongoose_component, #{pid => self()}),
-    mongoose_component:register_components(Routes, node(), Handler, AreHidden).
-
--spec get_routes(data()) -> [jid:lserver()].
-get_routes(#component_data{lserver = Subdomain, is_subdomain = true}) ->
-    StaticDomains = mongoose_domain_api:get_all_static(),
-    DynamicDomains = mongoose_domain_api:get_all_dynamic(),
-    [ component_route(Subdomain, Domain) || {Domain, _} <- StaticDomains ]
-    ++ [ component_route(Subdomain, Domain) || {Domain, _} <- DynamicDomains ];
-get_routes(#component_data{lserver = Host}) ->
-    [Host].
+    [ mongoose_domain_api:register_subdomain(HostType, Prefix, Handler)
+      || HostType <- ?ALL_HOST_TYPES ],
```

Comment:
I have two questions about it:
1. Don't we register the handler twice here: once for `SubDomain` (line 315), and once again for `HostType` (line 317)? Maybe it somehow works, but the resulting routing state would be messy.
2. Is the domain unregistered anywhere in the code? If it doesn't -- isn't it an issue that there will be leftover subdomains if the component disconnects?